lib/routes/account.js

```diff
@@ -385,7 +386,17 @@ module.exports = function (
               function (result) {
                 emailRecord = result
 
-                doSigninConfirmation = requestHelper.shouldEnableSigninConfirmation(emailRecord, config, request)
+                // Session token verification is only enabled for certain users during phased rollout.
+                // Even when it is enabled, we only do the email challenge if:
+                //  * the request wants keys, since unverified sessions are fine to use for e.g. oauth login.
+                //  * the email is verified, since content-server triggers a resend of the verification
+                //    email on unverified accounts, which doubles as sign-in confirmation.
+                if (! requestHelper.shouldEnableTokenVerification(emailRecord, config, request)) {
+                  tokenVerificationId = undefined
+                  doSigninConfirmation = false
+                } else {
+                  doSigninConfirmation = requestHelper.wantsKeys(request) && emailRecord.emailVerified
```

Comment:
You believe correctly, and we can try to add commentary to make that clearer in the code.